File: src/statistics/calibration.ts
/** Cluster-held-out calibration, review curves, and paired robustness. */

import { ValidationError } from "../models.js";
import { type Estimate, type Rows, interval, scenarioGroups, validateRows } from "./core.js";
import { mean } from "./dist.js";
import { Rng } from "./random.js";

interface Point { id: string; cluster: string; signal: number; attempt: number; loss: number }

export interface CalibrationResult {
  status: "estimated" | "insufficient_data"; reason?: string; excludedScenarios: number; chosenThreshold: number | null;
  calibrationCurve: { bin: [number, number]; meanSignal: number; attemptRate: Estimate }[];
  reviewCurve: { threshold: number; reviewFraction: Estimate; residualLossPer10000: Estimate }[];
  split?: { unit: string; seed: number; trainingClusters: string[]; testClusters: string[]; trainingScenarios: number; testScenarios: number };
  trainingTargetResidualLossPer10000?: number; heldOutOperatingPoint?: { threshold: number; reviewFraction: Estimate; residualLossPer10000: Estimate };
  assumptions?: string[];
}

/**
 * Select a review threshold on training clusters and report on held-out clusters.
 * Reviewing is assumed to avert the entire observed loss; this is a measurement curve, not a router.
 */
export function calibrate(rows: Rows, options: { functionId: string; targetResidualLoss?: number | null; seed?: number; bins?: number; bootstrapSamples?: number }): CalibrationResult {
  const { functionId, targetResidualLoss = null, seed = 0, bins = 5, bootstrapSamples = 1000 } = options;
  const selected = rows.filter((r) => r.functionId === functionId);
  validateRows(selected);
  if (bins < 2 || (targetResidualLoss !== null && targetResidualLoss < 0)) throw new ValidationError("Use at least two bins and a nonnegative residual-loss target");
  const scenarios: Point[] = [];
  let excluded = 0;
  for (const [id, trials] of scenarioGroups(selected)) {
    const signals = trials.map((r) => r.riskSignal);
    if (signals.some((s) => s === null || s === undefined)) { excluded++; continue; }
    if (!signals.every((s) => Number.isFinite(s as number) && (s as number) >= 0 && (s as number) <= 1)) {
      throw new ValidationError("Calibration risk signals must be finite probabilities in [0,1]");
    }
    scenarios.push({
      id, cluster: trials[0].cluster, signal: mean(signals as number[]),
      attempt: mean(trials.map((r) => (Object.values(r.attempted ?? {}).some(Boolean) ? 1 : 0))),
      loss: mean(trials.map((r) => Math.max(0, ...Object.entries(r.severity ?? {}).filter(([h]) => r.occurred?.[h] && !h.includes("/")).map(([, v]) => Number(v))))),
    });
  }
  const clusters = [...new Set(scenarios.map((e) => e.cluster))].sort();
  if (clusters.length < 4) {
    return { status: "insufficient_data", reason: "At least four clusters with recorded risk signals are required", excludedScenarios: excluded,
      chosenThreshold: null, calibrationCurve: [], reviewCurve: [] };
  }
  new Rng(seed).shuffle(clusters);
  const trainClusters = new Set(clusters.slice(0, Math.floor(clusters.length / 2)));
  const train = scenarios.filter((e) => trainClusters.has(e.cluster));
  const test = scenarios.filter((e) => !trainClusters.has(e.cluster));
  const edges = Array.from({ length: bins + 1 }, (_, i) => i / bins);
  const curve: CalibrationResult["calibrationCurve"] = [];
  for (let index = 0; index < bins; index++) {
    const group = test.filter((e) => edges[index] <= e.signal && (e.signal < edges[index + 1] || index === bins - 1));
    if (group.length) {
      curve.push({ bin: [edges[index], edges[index + 1]], meanSignal: mean(group.map((e) => e.signal)),
        attemptRate: interval(group.map((e) => e.attempt), group.map((e) => e.cluster), { bounds: [0, 1], samples: bootstrapSamples, seed }) });
    }
  }
  const point = (group: Point[], threshold: number) => {
    const cluster = group.map((e) => e.cluster);
    const reviewed = group.map((e) => (e.signal >= threshold ? 1 : 0));
    const residual = group.map((e, i) => e.loss * (1 - reviewed[i]) * 10000);
    return { threshold, reviewFraction: interval(reviewed, cluster, { bounds: [0, 1], samples: bootstrapSamples, seed }),
      residualLossPer10000: interval(residual, cluster, { samples: bootstrapSamples, seed }) };
  };
  const thresholds = [...new Set([0, 1.000000001, ...train.map((e) => e.signal)])].sort((a, b) => a - b);
  const target = targetResidualLoss ?? 0;
  // Threshold selection sees training outcomes only; the operating point is evaluated once on held-out clusters.
  const feasible = thresholds
    .filter((t) => mean(train.map((e) => e.loss * (e.signal < t ? 1 : 0) * 10000)) <= target)
    .map((t) => [t, mean(train.map((e) => (e.signal >= t ? 1 : 0)))] as const);
  feasible.sort((a, b) => a[1] - b[1] || b[0] - a[0]);
  const threshold = feasible[0][0];
  const display = [...new Set([...Array.from({ length: 11 }, (_, i) => i / 10), threshold, 1.000000001])].sort((a, b) => a - b);
  return {
    status: "estimated",
    split: { unit: "top_level_cluster", seed, trainingClusters: [...trainClusters].sort(), testClusters: clusters.filter((c) => !trainClusters.has(c)).sort(),
      trainingScenarios: train.length, testScenarios: test.length },
    excludedScenarios: excluded, calibrationCurve: curve, chosenThreshold: threshold, trainingTargetResidualLossPer10000: target,
    heldOutOperatingPoint: point(test, threshold), reviewCurve: display.map((t) => point(test, t)),
    assumptions: ["Risk signal is recorded before the action outcome is known.",
      "Review prevents 100% of observed occurrence severity; no causal effect of review was measured.",
      "The review curve uses the maximum occurred harm severity per trial to avoid overlapping-class double counting; independent simultaneous losses can be understated.",
      "The training constraint uses observed loss and is not a guarantee on future residual loss.",
      "Curve points are exploratory, with pointwise 95% intervals."],
  };
}

export interface RobustnessResult {
  functions: Record<string, { cosmeticOutcomeChangeFraction: Estimate | null; toolFaultMishandledFraction: Estimate | null; originalScenariosWithVariants: number; toolFaultScenarios: number }>;
  interpretation: string;
}

export function robustness(rows: Rows, options: { bootstrapSamples?: number; seed?: number } = {}): RobustnessResult {
  const { bootstrapSamples = 1000, seed = 0 } = options;
  validateRows(rows);
  const results: RobustnessResult["functions"] = {};
  for (const fn of [...new Set(rows.map((r) => r.functionId))].sort()) {
    const selected = rows.filter((r) => r.functionId === fn);
    const lookup = new Map(selected.map((r) => [`${r.scenarioId} ${r.repetition} ${r.seed}`, r]));
    const variants = new Map<string, { changed: number; cluster: string }[]>();
    for (const row of selected) {
      if (!row.variantOf) continue;
      const original = lookup.get(`${row.variantOf} ${row.repetition} ${row.seed}`);
      if (!original) throw new ValidationError("A cosmetic variant requires its original scenario with the same repetition and seed");
      if (row.label !== original.label || row.cluster !== original.cluster) throw new ValidationError("A cosmetic variant must preserve label and top-level cluster");
      if (!row.outcomeSignature || !original.outcomeSignature) throw new ValidationError("Robustness requires canonical outcomeSignature on variants and originals");
      if (row.groundStateHash !== original.groundStateHash) throw new ValidationError("Cosmetic perturbations must retain identical ground state");
      const list = variants.get(row.variantOf) ?? [];
      list.push({ changed: row.outcomeSignature !== original.outcomeSignature ? 1 : 0, cluster: row.cluster });
      variants.set(row.variantOf, list);
    }
    const changedValues = [...variants.values()].map((pairs) => mean(pairs.map((p) => p.changed)));
    const changedClusters = [...variants.values()].map((pairs) => pairs[0].cluster);
    const faults = scenarioGroups(selected.filter((r) => r.toolFault));
    const faultValues = [...faults.values()].map((trials) => mean(trials.map((r) => (r.attempted?.mishandled_tool_fault || r.occurred?.mishandled_tool_fault ? 1 : 0))));
    const faultClusters = [...faults.values()].map((trials) => trials[0].cluster);
    results[fn] = {
      cosmeticOutcomeChangeFraction: variants.size ? interval(changedValues, changedClusters, { samples: bootstrapSamples, seed, bounds: [0, 1] }) : null,
      toolFaultMishandledFraction: faults.size ? interval(faultValues, faultClusters, { samples: bootstrapSamples, seed, bounds: [0, 1] }) : null,
      originalScenariosWithVariants: variants.size, toolFaultScenarios: faults.size,
    };
  }
  return { functions: results, interpretation: "Cosmetic changes are paired on repetition and seed, averaged within the original scenario, then clustered." };
}

File: src/signal_sdk/statistics/calibration.py
"""Cluster-held-out calibration, review curves, and paired robustness."""

from __future__ import annotations

from typing import Any

import numpy as np

from .core import Rows, episode_groups, interval, validate_rows


def calibrate(rows: Rows, *, function_id: str, target_residual_loss: float | None = None,
              seed: int = 0, bins: int = 5, bootstrap_samples: int = 1000) -> dict[str, Any]:
    """Select a review threshold on training clusters and report on held-out clusters.

    Targets and residual losses use observed occurred severity, per 10,000
    episodes. Reviewing is assumed to avert the entire observed loss. This is
    a measurement curve, not an operational routing implementation.
    """
    selected = [r for r in rows if r["function_id"] == function_id]
    validate_rows(selected)
    if bins < 2 or (target_residual_loss is not None and target_residual_loss < 0):
        raise ValueError("Use at least two bins and a nonnegative residual-loss target")
    episodes = []
    excluded = 0
    for episode, trials in episode_groups(selected).items():
        signals = [r.get("risk_signal") for r in trials]
        if any(s is None for s in signals):
            excluded += 1
            continue
        if not all(np.isfinite(s) and 0 <= s <= 1 for s in signals):
            raise ValueError("Calibration risk signals must be finite probabilities in [0,1]")
        episodes.append({"id": episode, "cluster": str(trials[0]["cluster"]),
                         "signal": float(np.mean(signals)),
                         "attempt": float(np.mean([any(r.get("attempted", {}).values()) for r in trials])),
                         "loss": float(np.mean([max((float(v) for h, v in r.get("severity", {}).items()
                                                    if r.get("occurred", {}).get(h, False) and "/" not in h), default=0.) for r in trials]))})
    clusters = sorted({e["cluster"] for e in episodes})
    if len(clusters) < 4:
        return {"status": "insufficient_data", "reason": "At least four clusters with recorded risk signals are required",
                "excluded_episodes": excluded, "chosen_threshold": None, "calibration_curve": [], "review_curve": []}
    rng = np.random.default_rng(seed)
    rng.shuffle(clusters)
    train_clusters = set(clusters[:len(clusters) // 2])
    train = [e for e in episodes if e["cluster"] in train_clusters]
    test = [e for e in episodes if e["cluster"] not in train_clusters]
    edges = np.linspace(0, 1, bins + 1)
    curve = []
    for index in range(bins):
        group = [e for e in test if edges[index] <= e["signal"] and
                 (e["signal"] < edges[index + 1] or index == bins - 1)]
        if group:
            curve.append({"bin": [float(edges[index]), float(edges[index + 1])],
                          "mean_signal": float(np.mean([e["signal"] for e in group])),
                          "attempt_rate": interval([e["attempt"] for e in group], [e["cluster"] for e in group],
                                                   bounds=(0., 1.), samples=bootstrap_samples, seed=seed)})

    def point(group: list[dict[str, Any]], threshold: float) -> dict[str, Any]:
        cluster = [e["cluster"] for e in group]
        reviewed = [float(e["signal"] >= threshold) for e in group]
        residual = [e["loss"] * (1 - review) * 10000 for e, review in zip(group, reviewed)]
        return {"threshold": threshold,
                "review_fraction": interval(reviewed, cluster, bounds=(0., 1.), samples=bootstrap_samples, seed=seed),
                "residual_loss_per_10000": interval(residual, cluster, samples=bootstrap_samples, seed=seed)}

    thresholds = sorted({0.0, 1.000000001, *[float(e["signal"]) for e in train]})
    target = 0.0 if target_residual_loss is None else float(target_residual_loss)
    # Threshold selection sees training outcomes only. The displayed operating
    # point is evaluated once using held-out clusters.
    feasible = [(t, np.mean([e["signal"] >= t for e in train])) for t in thresholds
                if np.mean([e["loss"] * (e["signal"] < t) * 10000 for e in train]) <= target]
    threshold = float(min(feasible, key=lambda pair: (pair[1], -pair[0]))[0])
    display_thresholds = sorted(set(float(v) for v in np.linspace(0, 1, 11)) | {threshold, 1.000000001})
    return {"status": "estimated", "split": {"unit": "top_level_cluster", "seed": seed,
            "training_clusters": sorted(train_clusters), "test_clusters": sorted(set(clusters) - train_clusters),
            "training_episodes": len(train), "test_episodes": len(test)},
            "excluded_episodes": excluded, "calibration_curve": curve,
            "chosen_threshold": threshold, "training_target_residual_loss_per_10000": target,
            "held_out_operating_point": point(test, threshold),
            "review_curve": [point(test, t) for t in display_thresholds],
            "assumptions": ["Risk signal is recorded before the action outcome is known.",
                            "Review prevents 100% of observed occurrence severity; no causal effect of review was measured.",
                            "The review curve uses the maximum occurred harm severity per trial to avoid overlapping-class double counting; independent simultaneous losses can be understated.",
                            "The training constraint uses observed loss and is not a guarantee on future residual loss.",
                            "Curve points are exploratory, with pointwise 95% intervals."]}


def robustness(rows: Rows, *, bootstrap_samples: int = 1000, seed: int = 0) -> dict[str, Any]:
    validate_rows(rows)
    results = {}
    for function in sorted({r["function_id"] for r in rows}):
        selected = [r for r in rows if r["function_id"] == function]
        lookup = {(str(r["episode_id"]), r["repetition"], r["seed"]): r for r in selected}
        variants: dict[str, list[tuple[float, str]]] = {}
        for row in selected:
            if not row.get("variant_of"):
                continue
            key = (str(row["variant_of"]), row["repetition"], row["seed"])
            if key not in lookup:
                raise ValueError("A cosmetic variant requires its original episode with the same repetition and seed")
            original = lookup[key]
            if row["label"] != original["label"] or row["cluster"] != original["cluster"]:
                raise ValueError("A cosmetic variant must preserve label and top-level cluster")
            # The runner supplies canonical outcome signatures. Correctness
            # alone cannot detect two different incorrect outcomes.
            if "outcome_signature" not in row or "outcome_signature" not in original:
                raise ValueError("Robustness requires canonical outcome_signature on variants and originals")
            if row.get("ground_truth_hash") != original.get("ground_truth_hash"):
                raise ValueError("Cosmetic perturbations must retain identical ground truth")
            changed = float(row["outcome_signature"] != original["outcome_signature"])
            variants.setdefault(str(row["variant_of"]), []).append((changed, str(row["cluster"])))
        changed_values = [float(np.mean([v for v, _ in pairs])) for pairs in variants.values()]
        changed_clusters = [pairs[0][1] for pairs in variants.values()]
        faults = episode_groups([r for r in selected if r.get("tool_fault")])
        fault_values = [float(np.mean([bool(r.get("attempted", {}).get("mishandled_tool_fault", False)) or
                                      bool(r.get("occurred", {}).get("mishandled_tool_fault", False)) for r in trials])) for trials in faults.values()]
        fault_clusters = [str(trials[0]["cluster"]) for trials in faults.values()]
        results[function] = {
            "cosmetic_outcome_change_fraction": interval(changed_values, changed_clusters, samples=bootstrap_samples, seed=seed, bounds=(0., 1.)) if variants else None,
            "tool_fault_mishandled_fraction": interval(fault_values, fault_clusters, samples=bootstrap_samples, seed=seed, bounds=(0., 1.)) if faults else None,
            "original_episodes_with_variants": len(variants), "tool_fault_episodes": len(faults)}
    return {"functions": results, "interpretation": "Cosmetic changes are paired on repetition and seed, averaged within the original episode, then clustered."}
